Improvement: site event handling

TobiSW opened the pull request:
- Add SiteEvent
- Add SiteEvents
- Adjust Site save method to trigger pre and post save events
- Adjust Site delete method to trigger pre and post delete events

<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.5`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.5` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #13139 

## Additional info  
Here the new PR against 11.x. See previous PR #13140

TobiSW committed:
Improvement: SiteEvent

- Add SiteEvent
- Add SiteEvents constants
- Adjust Site save method to trigger pre and post save events

TobiSW committed:
Improvement: SiteEvent

- Adjust Site delete method to trigger pre and post delete events

TobiSW committed:
Merge branch 'pimcore:11.x' into site-event-handling

brusch commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,50 @@
+    {
+        $this->site = $site;
+    }
+
+    /**
+     * @return ElementInterface|Site
+     */
+    public function getElement()
+    {
+        return $this->getSite();
+    }
```
```suggestion
```
This is not needed, this is just intended to be used for event targets implementing `ElementInterface`.

brusch commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,50 @@
+use Pimcore\Model\Site;
+use Symfony\Contracts\EventDispatcher\Event;
+
+class SiteEvent extends Event implements ElementEventInterface
```
```suggestion
class SiteEvent extends Event
```

brusch commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,50 @@
+    use ArgumentsAwareTrait;
+
+    /** @var Site */
+    protected $site;
```
```suggestion
    protected Site $site;
```

brusch commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,50 @@
+        $this->arguments = $arguments;
+    }
+
+    /**
+     * @return Site
+     */
```
```suggestion
```

brusch commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,50 @@
+        return $this->site;
+    }
+
+    /**
+     * @param Site $site
+     */
```
```suggestion
```

brusch commented on models/Site.php:
```diff
@@ -470,4 +473,32 @@ public function getCreationDate()
+        $preSaveEvent = new SiteEvent($this);
+        $this->dispatchEvent($preSaveEvent, SiteEvents::PRE_SAVE);
+
+        $this->__call(__FUNCTION__, []);
```
```suggestion
        $this->getDao()->save();
```

brusch commented on models/Site.php:
```diff
@@ -470,4 +473,32 @@ public function getCreationDate()
+        $preDeleteEvent = new SiteEvent($this);
+        $this->dispatchEvent($preDeleteEvent, SiteEvents::PRE_DELETE);
+
+        $this->__call(__FUNCTION__, []);
```
```suggestion
        $this->getDao()->delete();
```
This is more accessible 😉

TobiSW replied:
done thx :+1:

TobiSW committed:
Improvement: SiteEvent

- Reformat code
- Remove not needed ElementEventInterface for SiteEvent class
- Make save and delete method in Site model more accessible

aashan10 commented:
@TobiSW the changes seem fine .. however phpstan is giving some errors, could you update your pr with thr latest changes from 11.x branch?

blankse commented on lib/Event/Model/SiteEvent.php:
```diff
@@ -0,0 +1,34 @@
+
+    protected Site $site;
+
+    /**
+     * @param Site $site
+     * @param array $arguments
+     */
```
```suggestion
```
We shouldn't set redundant php docs

TobiSW replied:
removed, thx

blankse commented on models/Site.php:
```diff
@@ -470,4 +473,32 @@ public function getCreationDate()
         return $this->creationDate;
     }
+
+    /**
+     * @return void
+     */
+    public function save()
```
```suggestion
    public function save(): void
```

blankse commented on models/Site.php:
```diff
@@ -470,4 +473,32 @@ public function getCreationDate()
+        $this->dispatchEvent($postSaveEvent, SiteEvents::POST_SAVE);
+    }
+
+    /**
+     * @return void
+     */
+    public function delete()
```
```suggestion
    public function delete(): void
```

TobiSW replied:
i kept it as it is in the complete model class to keep it consistent. I changed it now how you suggested it.

blankse replied:
We can't change it because it's a BC break. But in the 11.x release it will be changed, see https://github.com/pimcore/pimcore/issues/12240
For new methods we should set it instantly :)

TobiSW committed:
Merge branch 'pimcore:11.x' into site-event-handling

TobiSW committed:
Improvement: SiteEvent

- Remove redundant php docs

TobiSW committed:
Merge remote-tracking branch 'origin/site-event-handling' into site-event-handling

TobiSW committed:
Merge branch 'pimcore:11.x' into site-event-handling

aashan10 commented:
@dvesh3 could you please approve the workflow run?

dvesh3 commented:
> @dvesh3 could you please approve the workflow run?

done
